### PR TITLE
#985, #988, #33 Password reset etc.

### PR DESCRIPTION
**Pull request by squaregoldfish**
Added a Lost Password link to the login page, which emails a link to enter a new password.

Also:
* Improved the email texts (#33)
* Reformatted the email verification page (shown when clicking the link) (#988)
* Removed a load of unused code

### CODE REVIEW

**Comment by jonasfh** [comment]
Just some comments: 
1. I think it is wierd when links are not acting as links. This is not neccessarily a problem here, but I think it is strange for the user. I would recommend adding `<redirect/>` - statements in the faces-config file for the links in question
2. I would expect a redirect or a link back to the main page after the email is sent. In combination with 1., this makes it a little strange experience (from the url it looks like you are on the main page, but you would like to return to the main page.
3. Little curious why you are linking to reset_password.jsf (and why it works?) The actual file is called reset_password.xhtml. It also has the effect that you are afterwards redirected to index.jsf instead of index.xhtml (which also works!). Anyway, I think this should be changed to .xhtml for consistency.

File: WebApp/src/uk/ac/exeter/QuinCe/User/UserDB.java [hunk 1]
```diff
@@ -68,6 +68,11 @@
    */
   private static final String CLEAR_EMAIL_CODE_STATEMENT = "UPDATE user SET email_code = NULL, email_code_time = NULL WHERE email = ?";
 
+  /**
+   * SQL statement to remove a password reset code
+   */
+  private static final String CLEAR_PASSWORD_RESET_CODE_STATEMENT = "UPDATE user SET password_code = NULL, password_code_time = NULL WHERE email = ?";
```
Comment (jonasfh) [inline WebApp/src/uk/ac/exeter/QuinCe/User/UserDB.java]: Avoid long lines.

File: WebApp/src/uk/ac/exeter/QuinCe/User/UserDB.java [hunk 2]
```diff
@@ -532,53 +540,34 @@ public static int authenticate(DataSource dataSource, String email, char[] passw
...
+   * @param conn A database connection
    * @param user The user whose password is to be changed
-   * @param oldPassword The user's old password
    * @param newPassword The user's new password
-   * @return {@code true} if the password was changed successfully; {@code false} otherwise.
    * @throws DatabaseException If an error occurred
    * @throws MissingParamException If any of the parameters are null
    */
-  public static boolean changePassword(DataSource dataSource, User user, char[] oldPassword, char[] newPassword) throws DatabaseException, MissingParamException {
+  public static void changePassword(Connection conn, User user, char[] newPassword) throws DatabaseException, MissingParamException {
```
Comment (jonasfh) [inline WebApp/src/uk/ac/exeter/QuinCe/User/UserDB.java]: Avoid long lines.

File: WebApp/src/uk/ac/exeter/QuinCe/User/UserDB.java [hunk 3]
```diff
@@ -637,6 +626,31 @@ public static void clearEmailVerificationCode(DataSource dataSource, String emai
...
 
+  /**
+   * Remove the email verification code for a user. Note that we do not check whether
+   * the specified user exists.
+   * @param dataSource A data source
+   * @param email The user's email address
+   * @throws MissingParamException If any parameters are missing
+   * @throws DatabaseException If an error occurs while updating the database
+   */
+  public static void clearPasswordResetCode(Connection conn, String email) throws MissingParamException, DatabaseException {
```
Comment (jonasfh) [inline WebApp/src/uk/ac/exeter/QuinCe/User/UserDB.java]: Avoid long lines.

File: WebApp/src/uk/ac/exeter/QuinCe/jobs/user/SendCodeJob.java [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+  /**
+   * Job object constructor
+   * @param resourceManager The application's resource manager
+   * @param config The application configuration
+   * @param id The job ID
+   * @param params The job parameters
+   * @throws MissingParamException If any required parameters are missing
+   * @throws InvalidJobParametersException If the job parameters are invalid
+   */
+  public SendCodeJob(ResourceManager resourceManager, Properties config, long id, Map<String, String> params) throws MissingParamException, InvalidJobParametersException {
```
Comment (jonasfh) [inline WebApp/src/uk/ac/exeter/QuinCe/jobs/user/SendCodeJob.java]: Avoid long lines.

File: WebApp/src/uk/ac/exeter/QuinCe/jobs/user/SendPasswordResetMailJob.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+  /**
+   * Job object constructor
+   * @param resourceManager The application's resource manager
+   * @param config The application configuration
+   * @param id The job ID
+   * @param params The job parameters
+   * @throws MissingParamException If any required parameters are missing
+   * @throws InvalidJobParametersException If the job parameters are invalid
+   */
+  public SendPasswordResetMailJob(ResourceManager resourceManager, Properties config, long id, Map<String, String> params) throws MissingParamException, InvalidJobParametersException {
```
Comment (jonasfh) [inline WebApp/src/uk/ac/exeter/QuinCe/jobs/user/SendPasswordResetMailJob.java]: Avoid long lines.

File: WebApp/src/uk/ac/exeter/QuinCe/web/User/LoginBean.java [hunk 6]
```diff
@@ -138,6 +156,35 @@ public String signUp() {
...
+
+    User user = UserDB.getUser(getDataSource(), emailAddress);
+    if (null == user) {
+      setMessage(null, "The entered email address is not recognised");
+      result = null;
+    } else {
+      UserDB.generatePasswordResetCode(getDataSource(), user);
+      Map<String, String> emailJobParams = new HashMap<String, String>(1);
+      emailJobParams.put(SendPasswordResetMailJob.EMAIL_KEY, emailAddress);
+      JobManager.addInstantJob(ServletUtils.getResourceManager(), ServletUtils.getAppConfig(), user, "uk.ac.exeter.QuinCe.jobs.user.SendPasswordResetMailJob", emailJobParams);
```
Comment (jonasfh) [inline WebApp/src/uk/ac/exeter/QuinCe/web/User/LoginBean.java]: Avoid long lines.

File: WebApp/src/uk/ac/exeter/QuinCe/web/User/ResetPasswordBean.java [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+  private String password2 = null;
+
+  /**
+   * Verify the password reset code.
+   */
+  public void verifyCode() {
+
+    if (!verified) {
+      try {
+        int codeOK = UserDB.checkPasswordResetCode(ServletUtils.getDBDataSource(), getRequestParameter(USER_PARAM), getRequestParameter(CODE_PARAM));
```
Comment (jonasfh) [inline WebApp/src/uk/ac/exeter/QuinCe/web/User/ResetPasswordBean.java]: Avoid long lines.

**Comment by squaregoldfish** [comment]
Ready for another review

**Pull request merged**